File: packages/python/src/Seguranca/Ataques/AtaqueCoberturaDependenciaIV.py
from ..ResultadoAtaque import ResultadoAtaque
from ..SkipAtaqueException import SkipAtaqueException
from ..Util import random_int


class AtaqueCoberturaDependenciaIV:
    """
    Complemento do AtaqueCoberturaDependencia (que varia a CHAVE): aqui varia
    o IV. Toda posição do IV precisa influenciar toda posição de saída; uma
    posição "morta" do IV reduziria a entropia efetiva que o IV injeta.
    """

    def __init__(self, tamanho_bloco: int = 32, perturbacoes_por_posicao: int = 5):
        self.tamanho_bloco = tamanho_bloco
        self.perturbacoes_por_posicao = perturbacoes_por_posicao

    def nome(self) -> str:
        return "Cobertura de dependência do IV (entrada x saída)"

    def executar(self, alvo) -> ResultadoAtaque:
        chave = alvo.chave_de_teste()
        tamanho_iv = alvo.tamanho_iv()
        iv_base = b"\x00" * tamanho_iv

        ks_base = alvo.gerar_keystream_bruto(chave, iv_base, "enc", self.tamanho_bloco)
        if ks_base is None:
            raise SkipAtaqueException("Alvo não expõe gerar_keystream_bruto.")

        pares_independentes = []

        for pos_iv in range(tamanho_iv):
            afetou = [False] * self.tamanho_bloco

            for _ in range(self.perturbacoes_por_posicao):
                iv = bytearray(iv_base)
                iv[pos_iv] = random_int(1, 255)
                ks = alvo.gerar_keystream_bruto(chave, iv, "enc", self.tamanho_bloco)

                for pos_saida in range(self.tamanho_bloco):
                    if ks[pos_saida] != ks_base[pos_saida]:
                        afetou[pos_saida] = True

            for pos_saida, ok in enumerate(afetou):
                if not ok:
                    pares_independentes.append(f"iv[{pos_iv}] -> saida[{pos_saida}]")

        if pares_independentes:
            return ResultadoAtaque(
                self.nome(),
                True,
                "alta",
                f"{len(pares_independentes)} par(es) sem dependência detectável em "
                f"{self.perturbacoes_por_posicao} tentativas cada",
                {"pares": pares_independentes[:20]},
            )

        return ResultadoAtaque(
            self.nome(),
            False,
            "info",
            f"Todas as {tamanho_iv} posições do IV influenciam todas as "
            f"{self.tamanho_bloco} posições de saída",
        )

File: packages/php/src/Seguranca/Ataques/AtaqueCoberturaDependenciaIV.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\ResultadoAtaque;
use Application\Seguranca\SkipAtaqueException;

/**
 * Complemento do AtaqueCoberturaDependencia (que varia a CHAVE): aqui varia
 * o IV. Toda posição do IV precisa influenciar toda posição de saída; uma
 * posição "morta" do IV reduziria a entropia efetiva que o IV injeta.
 */
class AtaqueCoberturaDependenciaIV implements AtaqueInterface
{
    public function __construct(private int $tamanhoBloco = 32, private int $perturbacoesPorPosicao = 5)
    {
    }

    public function nome(): string
    {
        return 'Cobertura de dependência do IV (entrada x saída)';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $chave = $alvo->chaveDeTeste();
        $tamanhoIv = $alvo->tamanhoIv();
        $ivBase = str_repeat("\x00", $tamanhoIv);

        $ksBase = $alvo->gerarKeystreamBruto($chave, $ivBase, 'enc', $this->tamanhoBloco);
        if ($ksBase === null) {
            throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
        }

        $paresIndependentes = [];

        for ($posIv = 0; $posIv < $tamanhoIv; $posIv++) {
            $afetou = array_fill(0, $this->tamanhoBloco, false);

            for ($p = 0; $p < $this->perturbacoesPorPosicao; $p++) {
                $iv = $ivBase;
                $iv[$posIv] = chr(random_int(1, 255));
                $ks = $alvo->gerarKeystreamBruto($chave, $iv, 'enc', $this->tamanhoBloco);

                for ($posSaida = 0; $posSaida < $this->tamanhoBloco; $posSaida++) {
                    if ($ks[$posSaida] !== $ksBase[$posSaida]) {
                        $afetou[$posSaida] = true;
                    }
                }
            }

            foreach ($afetou as $posSaida => $ok) {
                if (!$ok) {
                    $paresIndependentes[] = "iv[$posIv] -> saida[$posSaida]";
                }
            }
        }

        if (!empty($paresIndependentes)) {
            return new ResultadoAtaque(
                $this->nome(),
                vulneravel: true,
                severidade: 'alta',
                detalhes: count($paresIndependentes) . " par(es) sem dependência detectável em {$this->perturbacoesPorPosicao} tentativas cada",
                dados: ['pares' => array_slice($paresIndependentes, 0, 20)],
            );
        }

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: false,
            severidade: 'info',
            detalhes: "Todas as $tamanhoIv posições do IV influenciam todas as {$this->tamanhoBloco} posições de saída",
        );
    }
}
